Fix example_value of map

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #124203


Previously, we didn't expand the shape of example_value of map to the same as inputs (edit: the first mapped dimension). This pr fixes this bug. To make this easier, we change _call_function_and_unflatten_output to accept example_values directly instead of retrieving them from the variable trackers. 

Also remove a redundant call function node in strict_mode higher order op in dynamo.

Test Plan:
existing tests.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

ydwu4 committed:
Fix example_value of map

[ghstack-poisoned]

ydwu4 committed:
Update on "Fix example_value of map"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ezyang commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -950,6 +968,13 @@ def call_function(
+            proxy.node.meta["example_value"] for proxy in body_r.as_proxy()
+        ]
+        map_example_out = [
+            t.expand(sample_shape[0], *t.shape) for t in subgraph_example_value
```
This seems sus, expand will generate a view aliasing with the subgraph example value, but you don't actually want to return a view right?

ydwu4 replied:
em..I just do a `torch.empty_strided`, does it sound like a good idea? BTW, what might go wrong if we return a view of it?

ydwu4 committed:
Update on "Fix example_value of map"



Previously, we didn't expand the shape of example_value of map to the same as inputs (edit: for the first dimension i.e. mapped dimension). This pr fixes this bug. To make this easier, we change _call_function_and_unflatten_output to accept example_values directly instead of retrieving them from the variable trackers. 

Also remove a redundant call function node in strict_mode higher order op in dynamo.

Test Plan:
existing tests.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ezyang commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -950,6 +968,22 @@ def call_function(
+            map_example_out = [
+                torch.empty_strided(
+                    size=(sample_shape[0], *t.size()),
+                    stride=(0, *t.stride()),
```
I'm still perplexed though. Is the output of map really expanded? This is hard for me to believe, because I thought you want to actually map the function on each row individually, so you couldn't possibly have an expanded output.

ydwu4 replied:
For map, we speculate the computation of one row because the map fn user that writes is written for one row programming-model wise. So we need to trace the subgraph using the sliced input and will re-use the subgraph for other rows. We assume other rows would produce the same output shapes and we just create an output based on single row's result. This would still be correct and can avoid specializing first dim.

ezyang replied:
In plain terms, if I say `ret.stride(0)` in eager, do I get zero or non-zero? Seems to me that I get a non-zero value.

ydwu4 replied:
Oh yeah! It should be mul(*t.sizes). I don't remember how did I compute the stride hhah.

ydwu4 committed:
Update on "Fix example_value of map"



Previously, we didn't expand the shape of example_value of map to the same as inputs (edit: for the first dimension to be accurate i.e. mapped dimension). This pr fixes this bug. To make this easier, we change _call_function_and_unflatten_output to accept example_values directly instead of retrieving them from the variable trackers. 

Also remove a redundant call function node in strict_mode higher order op in dynamo.

Test Plan:
existing tests.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -950,6 +968,32 @@ def call_function(
+                    *subgraph_out.stride(),
+                )
+
+        with tx.output.fake_mode:
+            map_example_out = [
+                torch.empty_strided(
+                    size=(sample_shape[0], *t.size()),
+                    stride=_compute_stride(t),
+                    dtype=t.dtype,
+                    layout=t.layout,
+                    device=t.device,
+                    requires_grad=t.requires_grad,
+                )
+                for t in subgraph_example_value
+            ]
```
If the semantics of map are "for-loop then stack", then I would expect the output tensor to be contiguous instead of having this stride. Also, can we add a test that the strides of map under FakeTensor are the same as the strides from eager?

ydwu4 replied:
yeah, stack does make the output contiguous. Added a test to compare eager run map_dense results has the same metadata as the `example_values`.

ydwu4 committed:
Update on "Fix example_value of map"



Previously, we didn't expand the shape of example_value of map to the same as inputs (edit: for the first dimension to be accurate i.e. mapped dimension). This pr fixes this bug. To make this easier, we change _call_function_and_unflatten_output to accept example_values directly instead of retrieving them from the variable trackers. 

Also remove a redundant call function node in strict_mode higher order op in dynamo.

Test Plan:
existing tests.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ydwu4 committed:
Update on "Fix example_value of map"



Previously, we didn't expand the shape of example_value of map to the same as inputs (edit: for the first dimension to be accurate i.e. mapped dimension). This pr fixes this bug. To make this easier, we change _call_function_and_unflatten_output to accept example_values directly instead of retrieving them from the variable trackers. 

Also remove a redundant call function node in strict_mode higher order op in dynamo.

Test Plan:
existing tests.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ezyang commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -950,6 +968,23 @@ def call_function(
+            # the same first dimension as the mapped input.
+            # We also do a clone with contiguous_format. This is to be consistent with
+            # eager semantic of map, which stacks the outputs. The result is contiguous
+            # as a result of the stack operation.
```
It's actually kind of surprising you don't try to preserve the inner layout (e.g., to preserve NHWC'ness, if necessary). I am not doubting that this is right, but maybe the eager semantics are not optimal either lol

ydwu4 replied:
yeah, we can probably address it in the future when we see a real use case.

ezyang commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -1292,6 +1292,47 @@ def forward(self, getitem, const):
+            torch.randn(3),
+            torch.randn(3, 4),
+            torch.randn(3, 4, 5, requires_grad=True),
+        ]
```
This seems very insufficient for testing strangely strided inputs?

ydwu4 replied:
added more strangely strided inputs

ydwu4 committed:
Update on "Fix example_value of map"



Previously, we didn't expand the shape of example_value of map to the same as inputs (edit: for the first dimension to be accurate i.e. mapped dimension). This pr fixes this bug. To make this easier, we change _call_function_and_unflatten_output to accept example_values directly instead of retrieving them from the variable trackers. 

Also remove a redundant call function node in strict_mode higher order op in dynamo.

Test Plan:
existing tests.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge
